apps/HelloAndroidCamera2/jni/native.cpp

```diff
@@ -0,0 +1,338 @@
+#include <jni.h>
+#include <android/log.h>
+#include <android/bitmap.h>
+#include <android/native_window_jni.h>
+
+#include <algorithm>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+#include "deinterleave.h"
+#include "edge_detect.h"
+#include "HalideRuntime.h"
+#include "HalideRuntimeOpenCL.h"
+
+#define LOGD(...) __android_log_print(ANDROID_LOG_DEBUG, "halide_native", __VA_ARGS__)
+#define LOGE(...) __android_log_print(ANDROID_LOG_ERROR, "halide_native", __VA_ARGS__)
+
+#define DEBUG 1
+#define IMAGE_FORMAT_YV12 842094169
+
+// Round x up to a multiple of mask.
+// E.g., ALIGN(x, 16) means round x up to the nearest multiple of 16.
+#define ALIGN(x, mask) (((x) + (mask)-1) & ~((mask)-1))
+
+extern "C" void halide_set_error_handler(int (*handler)(void *user_context, const char *));
```

Comment:
Aren't these declared in HalideRuntime.h? (If not, they should be. It's always preferable to use those decls rather than extern-ing them yourself here.)